### ISSUE TITLE
left.isCompileTimeValue is not a function' when build

### ISSUE THREAD

**Issue by Wuyialex** [storybookjs/addon-react-native-web#29]
After updating addon-react-native-web
 to 0.0.16 and installing other storybook modules to 6.4.3 (same version as in example), I tried to build, and the following problem occurred. However, regardless of the problem, there is no problem in using the storybook.

`
left.isCompileTimeValue is not a function
    at handleStrictEqualityComparison (/Users/alexgue/SourceCode/public/rn-next-storybook-web-monorepo/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:554:29)
`

Reproduction repo
https://github.com/Wuyialex/rn-next-storybook-web-monorepo/tree/fix/%40storybook%2Faddon-react-native-web%400.0.16

**Comment by dannyhw** [comment]
@Wuyialex thanks for reporting :), I will look into it soon.

With the holidays it might take longer to resolve but I'll let you know when I have updates about this.

**Comment by Wuyialex** [comment]
> @Wuyialex is this still happening? From what I can tell it's just a warning but I'll see if I can figure it out soon.

yes, it's still happening in reproduce repo.

**Comment by noeljackson** [comment]
I'm still experiencing this as well. @dannyhw Is it anything to be concerned about?

Where could I start to try and isolate the bug, at least to find where it is coming from?

Using version 0.0.18

**Comment by dannyhw** [comment]
To me it doesn't seem like a problem, seems to have no effect. Does it cause any problem for you?

**Comment by Wuyialex** [comment]
> To me it doesn't seem like a problem, seems to have no effect. Does it cause any problem for you?

not. It does not seem to cause problems such as crashes when using this.

**Comment by noeljackson** [comment]
No problems I can notice. I am curious, how did you isolate that the bug is coming from addon-react-native-web? I can't seem to trace the code back there very easily.

**Comment by dannyhw** [comment]
Personally I think its probably from another package, probably from the webpack5 builder just based on the error message in the issue description.

**Comment by Wuyialex** [comment]
> Personally I think its probably from another package, probably from the webpack5 builder just based on the error message in the issue description.

But this error not raised when 0.0.15. So It probably related with 0.0.16 upgrade.

**Comment by dannyhw** [comment]
But 0.0.16 doesn't have any actual changes, it was just documentation changes so probably you updated dependencies when running yarn add.

**Comment by dannyhw** [comment]
I'm going to close this for now, if its problematic for you then please let me know and I will reopen

**Comment by chrisdrackett** [comment]
seems like this is still an "issue" in as far as it fills the console with errors 🤔

**Comment by dannyhw** [comment]
@chrisdrackett this was closed since I wasn't able to find any reason why it would be related to this package specifically. So probably coming from a dependency which I don't control. Though if thats not the case I'm open to having another look.

**Comment by chrisdrackett** [comment]
I only ran into it after installing `addon-react-native-web` and it goes away if I remove it, ¯\_(ツ)_/¯ It does not seem to stop anything I use from working however.

**Comment by chrisdrackett** [comment]
@dannyhw ok, I now think this is more of an issue. I'm unable to use chromatic as the above makes it think that the build failed.

**Comment by dannyhw** [comment]
@chrisdrackett are you sure this is the error causing a failure? usually chromatic will only fail on an error, I thought this was just a warning?

**Comment by noeljackson** [comment]
@chrisdrackett @dannyhw The offending function for this seems to be fixed in the `webpack` dependency of this project. `webpack` @ v5.74.0 has a different `handleStrictEqualityComparison` function. 
https://github.com/webpack/webpack/blob/9fcaa243573005d6fdece9a3f8d89a0e8b399613/lib/javascript/JavascriptParser.js#L559

**Comment by chrisdrackett** [comment]
I get:

```
  ✖ Publish your built Storybook
    → Invalid Storybook build at /var/folders/8x/578wnsjs1z95jxlwvy8vdkbm0000gn/T/chromatic--35885-FGirz9CLmmmA (check the build log)
```

and the only thing in the log (other than normal progress) is

```
TypeError: left.isCompileTimeValue is not a function
    at handleStrictEqualityComparison (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:565:29)
    at /Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:790:13
    at Hook.eval (eval at create (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at JavascriptParser.evaluateExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:3215:25)
    at /Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:433:23
    at Hook.eval (eval at create (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at JavascriptParser.evaluateExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:3215:25)
    at /Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:433:23
    at Hook.eval (eval at create (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at JavascriptParser.evaluateExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:3215:25)
TypeError: left.isCompileTimeValue is not a function
    at handleStrictEqualityComparison (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:565:29)
    at /Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:790:13
    at Hook.eval (eval at create (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at JavascriptParser.evaluateExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:3215:25)
    at /Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:433:23
    at Hook.eval (eval at create (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at JavascriptParser.evaluateExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:3215:25)
    at /Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/ConstPlugin.js:247:30
    at Hook.eval (eval at create (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at JavascriptParser.walkLogicalExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:2494:55)
TypeError: left.isCompileTimeValue is not a function
    at handleStrictEqualityComparison (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:565:29)
    at /Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:790:13
    at Hook.eval (eval at create (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at JavascriptParser.evaluateExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:3215:25)
    at /Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/ConstPlugin.js:247:30
    at Hook.eval (eval at create (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at JavascriptParser.walkLogicalExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:2494:55)
    at JavascriptParser.walkExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:2299:10)
    at JavascriptParser.walkLeftRightExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:2483:8)
    at JavascriptParser.walkLogicalExpression (/Users/cwd/Developer/paradigm/node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavascriptParser.js:2496:9)
<s> [webpack.Progress] 53% building 11/14 entries 4799/4799 dependencies 1000/1026 modules
<s> [webpack.Progress] 57% building 12/14 entries 4799/4799 dependencies 1008/1026 modules
info => Manager built (8.02 s)

```

**Comment by noeljackson** [comment]
@chrisdrackett Your issue is with `@storybook/builder-webpack5` using v5 instead of v5.74.0 of `webpack` not with the `@storybookjs/addon-react-native-web` package. I suggest using `patch` to setup some fixes for this, and reporting your issue over here: https://github.com/storybookjs/storybook/

**Comment by chrisdrackett** [comment]
ok, I'll try and figure that out and open a new issue

**Comment by chrisdrackett** [comment]
@noeljackson hm, I'm trying with 5.74.0 and I'm getting the same thing. If I remove `'@storybook/addon-react-native-web',` from addons everything builds without errors (I'm not actually using any react-native-web components yet in my stories)

**Comment by noeljackson** [comment]
@chrisdrackett I was right and wrong. Essentially the problem is in webpack, but upgrading the version inside of @storybook/builder-webpack5 doesn't help.

What you can do is put a try catch statement around the offending code and it suppresses the errors.

If you patch @storybook/builder-webpack5's webpack version with patch-package, it works :)

This is an example of how I patched. For me the file was in: `node_modules/@storybook/builder-webpack5/node_modules/webpack/lib/javascript/JavaScriptParser.js`

```
	const handleStrictEqualityComparison = eql => {
					try {
					const left = this.evaluateExpression(expr.left);
					if (!left) return;
					const right = this.evaluateExpression(expr.right);
					if (!right) return;
					const res = new BasicEvaluatedExpression();
					res.setRange(expr.range);

					
					const leftConst = left.isCompileTimeValue();

					const rightConst = right.isCompileTimeValue();

					if (leftConst && rightConst) {
						return res
							.setBoolean(
								eql ===
									(left.asCompileTimeValue() === right.asCompileTimeValue())
							)
							.setSideEffects(
								left.couldHaveSideEffects() || right.couldHaveSideEffects()
							);
					}

					if (left.isArray() && right.isArray()) {
						return res
							.setBoolean(!eql)
							.setSideEffects(
								left.couldHaveSideEffects() || right.couldHaveSideEffects()
							);
					}
					if (left.isTemplateString() && right.isTemplateString()) {
						return handleTemplateStringCompare(left, right, res, eql);
					}

					const leftPrimitive = left.isPrimitiveType();
					const rightPrimitive = right.isPrimitiveType();

					if (
						// Primitive !== Object or
						// compile-time object types are never equal to something at runtime
						(leftPrimitive === false &&
							(leftConst || rightPrimitive === true)) ||
						(rightPrimitive === false &&
							(rightConst || leftPrimitive === true)) ||
						// Different nullish or boolish status also means not equal
						isAlwaysDifferent(left.asBool(), right.asBool()) ||
						isAlwaysDifferent(left.asNullish(), right.asNullish())
					) {
						return res
							.setBoolean(!eql)
							.setSideEffects(
								left.couldHaveSideEffects() || right.couldHaveSideEffects()
							);
					}
				} catch(e) {
					// SILENCE
				}
				};
```

**Comment by chrisdrackett** [comment]
ok, I'll give it a go. I guess I'm still curious what it is about this package that makes this error come up and weather its avoidable

**Comment by dannyhw** [comment]
@noeljackson @chrisdrackett thanks for the extra information. 

I have a few ideas of things to look into, I'll see if I can find anything out later today.

**Comment by chrisdrackett** [comment]
for now the patch (thanks @noeljackson) has unblocked me and I'm able to publish, but it would be nice to remove that crutch :D

**Comment by clayrisser** [comment]
This is still happening.

**Comment by dannyhw** [comment]
@clayrisser yes that's true though I don't really know what the cause is, and it seems to just spam the log rather than cause a particular issue.

I definitely want to fix this but can't dedicate time to debug it right now. Of course if it caused an blocking error I would raise the priority of it. 

However for now it's on my to do list and I'll get to it when I can. PR's very welcome if anyone figures it out in the meantime.
